src/visitable.cpp

```diff
@@ -212,6 +213,249 @@ VisitResponse visitable<vehicle_selector>::visit_items_with_parent(
+                // remove the entire stack
+                count -= stack->size();
+                res.splice( res.end(), *stack );
+                stack = inv->items.erase( stack );
+                if( count == 0 ) {
+                    return res;
+                }
+
+            } else {
+                // remove only some of the stack
+                char invlet = stack->front().invlet;
+                auto fin = stack->begin();
+                std::advance( fin, count );
+                res.splice( res.end(), *stack, stack->begin(), fin );
+                stack->front().invlet = invlet; // preserve invlet for remaining stacked items
+                return res;
+            }
+
+        } else {
+            // recurse through the contents of each stacked item separately
+            for( auto &e : *stack ) {
+                remove_internal( filter, e, count, std::back_inserter( res ) );
+                if( count == 0 ) {
+                    return res;
+                }
+            }
+
+            ++stack;
+        }
+    }
+    return res;
+}
+
+template <>
+std::list<item> visitable<Character>::remove_items_with( const
+        std::function<bool( const item &e )> &filter, int count )
+{
+    auto ch = static_cast<Character *>( this );
+    std::list<item> res;
+
+    // first try and remove items from the inventory
+    res = ch->inv.remove_items_with( filter, count );
+    count -= res.size();
+    if( count == 0 ) {
+        return res;
+    }
+
+    // then try any worn items
+    for( auto iter = ch->worn.begin(); iter != ch->worn.end(); ) {
+        if( filter( *iter ) ) {
+            res.splice( res.end(), ch->worn, iter++ );
+            if( --count == 0 ) {
+                return res;
+            }
+        } else {
+            remove_internal( filter, *iter, count, std::back_inserter( res ) );
+            if( count == 0 ) {
+                return res;
+            }
+            ++iter;
+        }
+    }
+
+    // finally try the currently wielded item (if any)
+    if( filter( ch->weapon ) ) {
+        res.push_back( ch->remove_weapon() );
+        count--;
+    } else {
+        remove_internal( filter, ch->weapon, count, std::back_inserter( res ) );
+    }
+
+    return res;
+}
+
+template <>
+std::list<item> visitable<map_cursor>::remove_items_with( const
+        std::function<bool( const item &e )> &filter, int count )
+{
+    auto cur = static_cast<map_cursor *>( this );
+    std::list<item> res;
+
+    if( count == 0 ) {
+        return res; // nothing to do
+    }
+
+    if( !g->m.inbounds( *cur ) ) {
+        debugmsg( "cannot remove items from map: cursor out-of-bounds" );
+        return res;
+    }
+
+    // fetch the appropriate item stack
+    int x, y;
+    submap *sub = g->m.get_submap_at( *cur, x, y );
+
+    for( auto iter = sub->itm[ x ][ y ].begin(); iter != sub->itm[ x ][ y ].end(); ) {
+        if( filter( *iter ) ) {
+            // check for presence in the active items cache
```

Comment:
Is there any reason why you don't use `map::i_rem` and instead copy all its content here? Same with getting the submap above, why not get a `map_item_stack` via `map::i_at`.